pkg/kubelet/cm/cgroup_manager_linux.go

```diff
@@ -557,3 +560,189 @@ func (m *cgroupManagerImpl) MemoryUsage(name CgroupName) (int64, error) {
+	}
+	mLim := int64(memLimit)
+	//TODO(vinaykul,InPlacePodVerticalScaling): Add memory request support
+	return &ResourceConfig{Memory: &mLim}, nil
+
+}
+
+// Get the resource config values applied to the cgroup for specified resource type
+func (m *cgroupManagerImpl) GetCgroupConfig(name CgroupName, resource v1.ResourceName) (*ResourceConfig, error) {
+	cgroupPaths := m.buildCgroupPaths(name)
+	cgroupResourcePath, found := cgroupPaths[string(resource)]
+	if !found {
+		return nil, fmt.Errorf("failed to build %v cgroup fs path for cgroup %v", resource, name)
+	}
+	switch resource {
+	case v1.ResourceCPU:
+		return getCgroupCpuConfig(cgroupResourcePath)
+	case v1.ResourceMemory:
+		return getCgroupMemoryConfig(cgroupResourcePath)
+	}
+	return nil, fmt.Errorf("unsupported resource %v for cgroup %v", resource, name)
+}
+
+func setCgroupv1CpuConfig(cgroupPath string, resourceConfig *ResourceConfig) error {
+	var cpuQuotaStr, cpuPeriodStr, cpuSharesStr string
+	if resourceConfig.CpuQuota != nil {
+		cpuQuotaStr = strconv.FormatInt(*resourceConfig.CpuQuota, 10)
+		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.cfs_quota_us"), []byte(cpuQuotaStr), 0700); err != nil {
+			return fmt.Errorf("failed to write %v to %v: %v", cpuQuotaStr, cgroupPath, err)
+		}
+	}
+	if resourceConfig.CpuPeriod != nil {
+		cpuPeriodStr = strconv.FormatUint(*resourceConfig.CpuPeriod, 10)
+		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.cfs_period_us"), []byte(cpuPeriodStr), 0700); err != nil {
+			return fmt.Errorf("failed to write %v to %v: %v", cpuPeriodStr, cgroupPath, err)
+		}
+	}
+	if resourceConfig.CpuShares != nil {
+		cpuSharesStr = strconv.FormatUint(*resourceConfig.CpuShares, 10)
+		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.shares"), []byte(cpuSharesStr), 0700); err != nil {
+			return fmt.Errorf("failed to write %v to %v: %v", cpuSharesStr, cgroupPath, err)
+		}
+	}
+	return nil
+}
+
+func setCgroupv2CpuConfig(cgroupPath string, resourceConfig *ResourceConfig) error {
+	if resourceConfig.CpuQuota != nil {
+		if resourceConfig.CpuPeriod == nil {
+			return fmt.Errorf("CpuPeriod must be specified in order to set CpuLimit")
+		}
+		cpuLimitStr := Cgroup2MaxCpuLimit
+		if *resourceConfig.CpuQuota > -1 {
+			cpuLimitStr = strconv.FormatInt(*resourceConfig.CpuQuota, 10)
+		}
+		cpuPeriodStr := strconv.FormatUint(*resourceConfig.CpuPeriod, 10)
+		cpuMaxStr := fmt.Sprintf("%s %s", cpuLimitStr, cpuPeriodStr)
+		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.max"), []byte(cpuMaxStr), 0700); err != nil {
+			return fmt.Errorf("failed to write %v to %v: %v", cpuMaxStr, cgroupPath, err)
+		}
+	}
+	if resourceConfig.CpuShares != nil {
+		cpuWeight := CpuSharesToCpuWeight(*resourceConfig.CpuShares)
+		cpuWeightStr := strconv.FormatUint(cpuWeight, 10)
+		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.weight"), []byte(cpuWeightStr), 0700); err != nil {
```

Comment:
Apologies if this was already discussed, but it appears this is used to resize the pod cgroup. 

If running with systemd cgroup driver, (default on cgroupv2) the pod cgroup would be managed by a systemd slice and kubelet shouldn't directly write to the cgroupfs. Instead, we should go through libcontainer which will update the slice in systemd (i.e. this path - https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/systemd/v2.go#L443-L457)

Otherwise, systemd slice and the cgroup values will differ and I believe a `systemctl daemon-reload` would reset the cgroup values back to what systemd slice is configured with, reverting the resize changes. 

This is something we can address as followup, not a blocker for alpha, but something we would definitely need to fix.

Reply:
Thanks for identifying this. We need to fix this. I'll create an issue to track this.

Reply:
Tracking issue https://github.com/kubernetes/kubernetes/issues/113857